04_annotation_Across_Samples_exploration (#776)
AlexsLemonade/OpenScPCA-analysis: analyses/cell-type-wilms-tumor-06/notebook/04_annotation_Across_Samples_exploration.Rmd

---
title: "Annotation exploration for `r params$scpca_project_id`"
author: "Maud PLASCHKA"

> date: '2024-08-13'

Reviewer: If you use this instead, it will automatically update when knit to html:
```suggestion
date: "`r Sys.Date()`"
```

> params:
>   scpca_project_id: "SCPCP000006"
>   sample_id: "SCPCS000176"
>   seed: 12345

Reviewer: We don't need any of these for this notebook actually, so they can be removed

```suggestion
```

output: 
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
    highlight: pygments
    df_print: paged
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE,
                      message=FALSE,
                      warnings=FALSE
                      )

```

## Introduction

> The aim is to explore the annotations and label transfers for all of the samples in `r params$scpca_project_id`.

Reviewer: For this module, it will always be the same project :) 
```suggestion
The aim is to explore the annotations and label transfers for all of the samples in SCPCP000006.
```

In order to explore the label transfer results, we look into some marker genes, table and percentages of cells in each annotation groups (from label transfers).

### Packages

```{r packages, message=FALSE, warning=FALSE}
library(Seurat)
library(tidyverse)
library(patchwork)

> set.seed(params$seed)

Reviewer: ```suggestion
```

metadata <- read.table(sample_metadata_file, sep = "\t", header = TRUE)

# Create a data frames of all annotations
cell_type_df <- metadata$scpca_sample_id |>
  purrr::map(
    # For each sample_id, do the following:
    \(sample_id) {

      # Read in the Seurat object
      srat <- readRDS(
        file.path(result_dir, sample_id, paste0("02b-fetal_kidney_label-transfer_", sample_id, ".Rds"))
      )

      # Create and return a data frame from the Seurat object with relevant annotations
      # this data frame will have four columns: barcode, sample_id, compartment, organ
      data.frame(
        compartment = srat$fetal_kidney_predicted.compartment,
        organ = srat$fetal_full_predicted.organ, 
        PTPRC = FetchData(object = srat, vars = "ENSG00000081237", layer = "counts"),
        VWF = FetchData(object = srat, vars = "ENSG00000110799", layer = "counts")
        ) |>
        tibble::rownames_to_column("barcode") |>
        dplyr::mutate(sample_id = sample_id)
    }
  ) |>
  # now combine all dataframes to make one big one
  dplyr::bind_rows()

```

### Output file

The report will be saved in the `notebook` directory. 

## Functions

> #### do_Table_Heatmap

Reviewer: You can delete this function, based on my other suggestions.

  
  df <- cell_type_df %>%
  group_by( !!sym(first_group), !!sym(last_group))%>% 
  summarise(Nb = n())  %>%
  mutate(C = sum(Nb)) %>%
  mutate(percent = Nb/C*100)

  
p1 <- ggplot(df, aes(x= !!sym(first_group), y = !!sym(last_group), fill = Nb)) +
   geom_tile() +
   scale_fill_viridis_c()  +
   theme_bw() +
   theme(text = element_text(size = 20)) +
   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
   guides(fill=guide_legend(title=" Number of \n cells"))

p2 <- ggplot(df, aes(x= !!sym(first_group), y = !!sym(last_group), fill = percent)) +
   geom_tile() +
   scale_fill_viridis_c()  +
   theme_bw() +
   theme(text = element_text(size = 20)) +
   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
   guides(fill=guide_legend(title=" Percentage of \n cells"))

p <- p1 + p2 + plot_layout(ncol = 2)
return(p)
}
```

#### do_Feature_mean

`do_Feature_mean` shows heatmap of mean expression of a feature grouped by a metadata. 

- `group.by` is the name of the metadata to group the cells
- `feature` is the name of the gene to average the expression

```{r fig.width=10, fig.height=4, out.width='100%'}

> do_Feature_mean <- function(group.by, feature ){

Reviewer: If we end up keeping this function, it should also definitely take the data frame (`cell_type_df`) as a formal argument and not just assume it's defined.

  
  df <- cell_type_df %>%
  group_by(sample_id, !!sym(group.by))%>% 
  summarise(m = mean(!!sym(feature)))  

  
p <- ggplot(df, aes(x= sample_id, y = !!sym(group.by), fill = m)) +
   geom_tile() +
   scale_fill_viridis_c()  +
   theme_bw() +
   theme(text = element_text(size = 20)) +
   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
   guides(fill=guide_legend(title=paste0(feature)))

return(p)
}
```

## Analysis

### Total number of cells per samples

```{r fig.width=5, fig.height=10, out.width='100%', results='asis'}
tmp <- table(cell_type_df$sample_id)
knitr::kable(tmp)
```
### Predicted organ

The predicted organ comes from label transfer from the human fetal atlas. 
This is the reference developed by [Cao et al.](https://www.science.org/doi/10.1126/science.aba7721) provided by `Azimuth` for label transfer.
The reference contain cells from 15 organs including kidney from fetal samples.
The label transfer have been performed in the notebook `02a_fetal_full_reference_Cao_{sample_id}.Rmd`

> The fetal full reference (Cao et al.) provides three levels of annotations:
> 
> - `fetal_full_predicted.organ` is one of the 15 organs used for building the reference. 
> We were nicely surprised that, for many of the samples, most of the cells from the Wilms tumor cohort mapped to fetal kidney cells. 
> 
> - `fetal_full_predicted.annotation.l1` holds predicted cell type annotation
> 
> - `fetal_full_predicted.annotation.l2` holds the combination of the predicted organ and predicted cell type annotation

Reviewer: Make this specific to the notebook:

```suggestion
This notebook looks at the `fetal_full_predicted.organ` annotation from this reference. 
```

    

```{r fig.width=25, fig.height=6,  results='asis'}
tmp <- table(cell_type_df$sample_id, cell_type_df$organ)

knitr::kable(tmp)
```

```{r fig.width=25, fig.height=6}
do_Table_Heatmap(first_group = "sample_id", last_group = "organ")

> ```

Reviewer: We can delete these chunks and replace with this code to make a table of counts for kidney vs not kidney. Then, style the data table however you usually like to do!

This table will have 3 columns in addition to the sample id: a total cell count, the percentage of cells labeled kidney, the percentage of cells labeled anything else that isn't kidney. This will make it easy for us to see which samples do really well, and which don't, for this reference.

```
kidney_count_df <- cell_type_df |>
  # make a new variable that tells us if it's kidney or not
  dplyr::mutate(organ_type = ifelse(organ == "Kidney", "kidney", "not_kidney"))  |>
  # count how many kidney or not
  dplyr::count(sample_id, organ_type) |>
  # make the data frame wide and use a value of 0 if a count is missing
  tidyr::pivot_wider(names_from = organ_type, values_from = n, values_fill=n) |>
  # add a column for total, and convert other counts to percentages
  dplyr::mutate(
    total = kidney + not_kidney, 
    kidney = round(kidney/total, 4)*100, 
    not_kidney = round(not_kidney/total, 4)*100) |>
  # arrange in descending order of percentage of kidney cells
  dplyr::arrange(desc(kidney)) 

DT::datatable(kidney_count_df)
```

Author: This is great thank you. I didn't know the function `pivot_wider`, that is great!
It is now running with `values_fill=0` instead of `values_fill=n`, if I understood correctly the help of the function and your code :)

### Predicted compartment

The predicted compartment is the result of the label transfer fron the human fetal kidney atlas: 
[Stewart et al.](https://www.science.org/doi/10.1126/science.aat5031) created a [human fetal kidney atlas](https://www.kidneycellatlas.org/). 
This reference contains only fetal kidney cells and has been precisely annotated by kidney experts. 
The label transfer have been performed in the notebook `02b_fetal_kidney_reference_Stewart_{sample_id}.Rmd`

The fetal kidney reference (Stewart et al.) provides two levels of annotations:

- `fetal_kidney_predicted.compartment` is one of the 4 main compartments composing a fetal kidney.
We expect immune and endothelial cells to be healthy (non-canerous) cells identified easily with high confidency. 
We expect stroma and fetal nephron compartment to contain both normal and malignant cells.

    
    - `immune` cells
    
    - `stroma` cells
    
    - `fetal_nephron` cells
    
    - `endothelial` cells
    

> - `fetal_kidney_predicted.cell_type` holds predicted cell type annotations that details further the four compartments, as for example:
> 
>     - `immune` subtypes: myeloid cells, lymphocytes, ...
>     
>     - `stroma` subtypes: mesenchymal stem cells, fibroblasts, ...
>     
>     - `fetal_nephron` subtypes: podocytes, uteric bud, mesenchymal cells (that appear to be cap mesenchyme cells) ...
>     
>     - `endothelial` cells
> 
> 
> We think that the first level of annotation `fetal_kidney_predicted.compartment` can be useful in identifying non-cancer cells, at least the endothelial and immune cells. 

Reviewer: Can be removed for this notebook

```suggestion
```

```{r fig.width=25, fig.height=3, results='asis'}
tmp <- table(cell_type_df$sample_id, cell_type_df$compartment)

knitr::kable(tmp)
```

```{r fig.width=25, fig.height=3}

do_Table_Heatmap(first_group = "sample_id", last_group = "compartment")

> ```

Reviewer: Replace these with the following and again style the DT however you like!

This table will show counts for each of the cell types, and again will make it easy for us to see which samples seem to have a decent population of cells we think are normal.

```
compartment_df <- cell_type_df |>
  # count how many per compartment
  dplyr::count(sample_id, compartment) |>
  # make the table wide and use a value of 0 if a count is missing
  tidyr::pivot_wider(names_from = compartment, values_from = n, values_fill = 0) |>
  # order the columns how we want to show them
  dplyr::select(sample_id, fetal_nephron, stroma, endothelium, immune) 

DT::datatable(compartment_df)
```

### Evaluate annotations with marker genes

Here we evaluate with marker genes the identification of endothelial, immune cells sub-populations.

##### Endothelial cells

We look at the endothelial marker "ENSG00000110799" = "VWF"

> ```{r fig.height=3, fig.width=15, message=FALSE, warning=FALSE, out.width='100%'}

Reviewer: This needs to be a lot taller to be able to see. I can't directly suggest on this line with GitHub, but try like 6 inches?

do_Feature_mean(  feature = "ENSG00000110799",
                  group.by = "compartment")
```

> 

Reviewer: Let's also try a boxplot and compare to see which kind of plot we think is more informative. What do you think? I wrote this code to start:

```
ggplot(cell_type_df, 
    aes(x = compartment, y = ENSG00000081237, fill = compartment)
  ) + 
  geom_boxplot(size = 0.5, size.outlier = 0.25) +
  facet_wrap(vars(sample_id), scale = "free_y") + 
  theme_bw() +
  theme(
    axis.text.x = element_text(size = rel(0.75), angle = 30), 
    legend.position = "none"
  )
```

##### Immune cells

We look at the immune marker "ENSG00000081237" = "PTPRC" alias "CD45"

> ```{r fig.height=3, fig.width=15, message=FALSE, warning=FALSE, out.width='100%'}

Reviewer: Same here, try 6 inches?

DT::datatable(compartment_df, 
              caption = "counts of cells in each compartment",
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',
                               buttons = c( 'csv', 'excel')))
```

We also checked the number of cell in each compartment per sample, to assess the presence/absence of non-cancer cells (endothelia and immune) that could help the inference of copy number alterations. 

```{r fig.width=25, fig.height=3, results='asis'}
compartment_df <- cell_type_df |>
  # count how many per compartment
  dplyr::count(sample_id, compartment) |>
  # make the table wide and use a value of 0 if a count is missing
  tidyr::pivot_wider(names_from = compartment, values_from = n, values_fill = 0) |>
  # order the columns how we want to show them
  dplyr::select(sample_id, fetal_nephron, stroma, endothelium, immune) 

DT::datatable(compartment_df, 
              caption = "counts of cells in each compartment",
              extensions = 'Buttons', 
              options = list(  dom = 'Bfrtip',
                               buttons = c( 'csv', 'excel')))
```

### Evaluate annotations with marker genes

> Here we evaluate with marker genes the identification of endothelial, immune cells sub-populations.

Reviewer: Let's also add a bit more:

```suggestion
Here we evaluate with marker genes the identification of endothelial, immune cells sub-populations.
We expect that cells labeled as endothelial and immune cells will have higher expression of these markers compared to other cell types. 
```
